pkg/machines/xmlCreator.js

```diff
@@ -85,6 +85,7 @@ export function getNetworkXML({ name, forwardMode, device, ipv4, netmask, ipv6,
         forwardMode === 'open') {
         const domainElem = doc.createElement('domain');
         domainElem.setAttribute('name', name);
+        domainElem.setAttribute('localOnly', 'yes');
```

Comment:
From libvirt documentation:
`If the optional localOnly attribute on the domain element is "yes", then DNS requests under this domain will only be resolved by the virtual network's own DNS server - they will not be forwarded to the host's upstream DNS server. If localOnly is "no", and by default, unresolved requests will be forwarded.`

Wouldn't this resolve only in limiting us? Since we would be only be limited on virtual network's DNS for domain translation, and if no entry is found there, then request is not forwarded? Correct me if I'm wrong, but wouldn't this result in lost of domain translation for Internet?

Reply:
The network name is used by the virtual DNS, for example "default", "local", "intranet" and most probably not like "google.com". By default we have the "default" network & domain, so any "vm" would be registered as "vm.default" in the virtual DNS. Request for "vm.default" and "inexistent.vm" would be dealt with by the virtual DNS  without external traffic, and "www.google.com" would be also resolved by the virtual DNS by forwarding upstream the query.
In my setup, I have my own DNS server configured to forward the "default" domain requests to the virtual DNS, so it can resolve on my network both internet and vm names. In case the "localOnly" option is not used, there is a loop between the DNS server and the virtual DNS for trying to resolve the "inexistentvm.default" or the "stoppedvm.default", because the virtual DNS will forward all not known requests to the upstream DNS which in this case if my main DNS server that will send the request back to the virtual DNS since it's related to "default" domain.
Once the "localOnly" option is used, then the virtual DNS will know that is the only DNS capable to solve hosts from "default" domain (existent or not), any other domains would be forwarded upstream for resolution.
I mentioned in the commit details the other scenario if we want to have network names like "google.com", which looks to be a nice hack, but is it a desired feature to have? if yes, then might be an idea to have a checkbox for using "localOnly".
(any time it's possible to add it by editing the network outside the cockpit interface, not handy for everybody)

Reply:
> In my setup, I have my own DNS server configured to forward the "default" domain requests to the virtual DNS

But this needs to be configured extra, right? And would this still work even if the user doesn't have their DNS server set up that way? It's not good to introduce "localOnly=yes" option as a default setting if it needs some additional configuration to make it work.
Maybe allowing user to optionally change virtual network as localOnly should be better solution.

Reply:
> I mentioned in the commit details the other scenario if we want to have network names like "google.com", which looks to be a nice hack, but is it a desired feature to have?

I don't think it's big use case. I have not seen any user wishing to use virtual networks on Cockpit (or libvirt) for this.

Reply:
No extra configuration is needed, I only did that because I wanted it to. Usual requirement is to have the VM to resolve internet names and other VM names from same virtual network, both of which are not affected negatively by the 'localOnly=yes'.
In fact it improves resolution of non-existing or stopped VM names.

Reply:
> No extra configuration is needed, I only did that because I wanted it to. Usual requirement is to have the VM to resolve internet names and other VM names from same virtual network, both of which are not affected negatively by the 'localOnly=yes'.

I think you need to either configure network so that in DHCP it tells VMs to use DNS of the resolving VM, or configure VMs statically. And the resolving VM has to have two interfaces, right? One connected to the network for which it's doing the DNS translation and the other connected to the outside world (e.g. a macvtap or another network).

> In fact it improves resolution of non-existing or stopped VM names.

Yeah, I can imagine that this offers more options to play with than what libvirt provides. But stopped VMs will also be translated by libvirt network as long as the DHCP lease doesn't time out.

localOnly is what you want, sure. But only for this narrow use case. You definitely do not want to turn it on by default for all networks and IIUC that is what Simon is objecting to.

Reply:
> I think you need to either configure network so that in DHCP it tells VMs to use DNS of the resolving VM, or configure VMs statically. And the resolving VM has to have two interfaces, right? One connected to the network for which it's doing the DNS translation and the other connected to the outside world (e.g. a macvtap or another network).

No, I think there is a confusion between "local only network" (isolated) and "local only domain". There is no need to have 2 interfaces, because "localOnly" is not influencing the traffic rules, the network type (ex. NAT) would still function as expected.
The "localOnly" option refers only to the way how DNS will handle requests for the specified domain name, for example in case of "default" network the domain name is "default" which would would mean that the following line:
`local=/default/`
would be added in the `/var/lib/libvirt/dnsmasq/default.conf` file,
Based on `dnsmasq` config documentation means:
`    # Add local-only domains here, queries in these domains are answered`
`    # from /etc/hosts or DHCP only.`
`    #local=/localnet/`

Exactly, name resolution is OK until DHCP leash expires, then name resolution would encounter the mentioned problem.

My scenario might look "narrow use case" it is just a scenario in which the impact is more visible so a root cause analysis was the way to go, but the problem of name resolution is not specific.

The simplest use case that can reproduce the problem is to have 2 VMs `vm1` and `vm2` that should resolve each others name, after a host restart boot just one of them `vm1` (just imagine that vm2 failed to start or that it crashed and lease expired) and try to resolve `vm2` name while monitoring the DNS queries on the host with `tcpdump -i any -n port 53 | grep vm2`. 
In the output there will be queries to the Virtual DNS and to the upstream the DNS configured on the host (one of: router DNS, ISP DNS, Google DNS etc.)
Using the "localOnly=yes" would eliminate from that chain the upstream DNS configured on the host, ie. `vm2` and `vm2.default` would results in immediate reply from the virtual DNS regardless of the state of `vm2`.
Everything else would work in the same way.

My questions are:
 - is it OK to have VM names "leak" to ISP/Internet ? unnecessary traffic, small but why not eliminate it if not needed.
 - is it common to resolve "somename" in a VM and expect it to be handled by upstream DNS to "somename.site.com" or "somename" on the physical network? to reach internet full names are generally used, short names are expected to be names on local network (virtual network in this case).